### PR TITLE
Add vm.format to call __format__ like PyObject_Format

### PR DESCRIPTION
**Pull request by youknowone**

### CODE REVIEW

File: vm/src/builtins/object.rs [hunk 1]
```diff
@@ -242,15 +242,18 @@ impl PyBaseObject {
...
 
     #[pymethod(magic)]
-    fn format(obj: PyObjectRef, format_spec: PyStrRef, vm: &VirtualMachine) -> PyResult<PyStrRef> {
-        if format_spec.as_str().is_empty() {
-            obj.str(vm)
-        } else {
-            Err(vm.new_type_error(format!(
+    fn format(
+        obj: PyObjectRef,
+        format_spec: Option<PyStrRef>,
```
Comment (youknowone) [inline vm/src/builtins/object.rs:247]: this is optional

File: vm/src/builtins/str.rs [hunk 2]
```diff
@@ -745,10 +745,24 @@ impl PyStr {
...
+        vm: &VirtualMachine,
+    ) -> PyResult<PyStrRef> {
+        let spec = spec.as_ref().map_or("", |spec| spec.as_str());
+        if spec.is_empty() {
+            return if zelf.class().is(vm.ctx.types.str_type) {
+                Ok(zelf.to_owned())
+            } else {
+                zelf.as_object().str(vm)
+            };
+        }
```
Comment (youknowone) [inline vm/src/builtins/str.rs:760]: str optimization when no spec

File: vm/src/protocol/object.rs [hunk 3]
```diff
@@ -330,12 +330,22 @@ impl PyObject {
...
-            s.try_into_value(vm)
-        }
+        let obj = match self.to_owned().downcast_exact::<PyStr>(vm) {
+            Ok(s) => return Ok(s.into_pyref()),
+            Err(obj) => obj,
+        };
+        // TODO: replace to obj.class().slots.str
+        let str_method = match vm.get_special_method(obj, identifier!(vm, __str__))? {
+            Ok(str_method) => str_method,
+            Err(obj) => return obj.repr(vm),
```
Comment (youknowone) [inline vm/src/protocol/object.rs:340]: repr is called here. it means we may have duplicated repr call somewhere.

File: vm/src/builtins/object.rs [hunk 4]
```diff
@@ -183,6 +183,7 @@ impl PyBaseObject {
     /// Return str(self).
     #[pymethod(magic)]
     fn str(zelf: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyStrRef> {
+        // FIXME: try tp_repr first and fallback to object.__repr__
```
Comment (youknowone) [inline vm/src/builtins/object.rs:186]: ah, found the redundant place. but no way to fix it before tp_repr

**Review by DimitrisJim — APPROVED** [review]
I only quickly glanced through the CPython equivalent but this seems good to me

**Pull request merged**